Hermes/appscale/hermes/stats/subscribers/profile.py

```diff
@@ -47,30 +55,380 @@ def receive(self, nodes_stats_dict):
+    for appending new lines.
+
+    Args:
+      node_ip: a string representation of node IP
+      monit_name: a string name of process as it's shown in monit status
+    Returns:
+      a file object opened for appending new data
+    """
+    node_dir = path.join(self._directory, node_ip)
+    file_name = path.join(node_dir, '{}.csv'.format(monit_name))
+    if not path.isfile(file_name):
+      try:
+        # Ensure directory for processes stats from the node exists
+        makedirs(node_dir)
+      except OSError as os_error:
+        if os_error.errno == errno.EEXIST and path.isdir(node_dir):
+          pass
+        else:
+          raise
+      # Create file and write header
+      with open(file_name, 'w') as csv_file:
+        csv.writer(csv_file).writerow(self._header)
+    # Open file for appending new data
+    return open(file_name, 'a')
+
+  def _get_summary_columns(self):
+    """ Opens summary-cpu-time.csv file (other summary file would be fine)
+    and reads it's header. Profiler needs to know order of columns previously
+    written to the summary.
+
+    Returns:
+      a list of column names: ['utc_timestamp', <service1>, <service2>, ..]
+    """
+    cpu_summary_file_name = self._get_summary_file_name('cpu_time')
+    if not path.isfile(cpu_summary_file_name):
+      return ['utc_timestamp']
+    with open(cpu_summary_file_name, 'r') as summary_file:
+      reader = csv.reader(summary_file)
+      return reader.next()  # First line is a header
+
+  def _save_summary(self, services_summary):
+    """ Saves services summary for each resource (cpu, resident memory and
+    unique memory). Output is 3 files (one for each resource) which
+    have a column for each service + utc_timestamp column.
+
+    Args:
+      services_summary: a dict where key is name of service and value is
+        an instance of ServiceProcessesSummary.
+    """
+    old_summary_columns = self._get_summary_columns()
+
+    for attribute in attr.fields(self.ServiceProcessesSummary):
+      # For each kind of resource (cpu, resident_mem, unique_mem)
+
+      summary_file_name = self._get_summary_file_name(attribute.name)
+
+      if len(old_summary_columns) == 1:
+        # Summary wasn't written yet - write header line to summary file
+        with open(summary_file_name, 'w') as new_summary:
+            csv.writer(new_summary).writerow(self._summary_columns)
+
+      if len(old_summary_columns) < len(self._summary_columns):
+        # Header need to be updated - add new services columns
+        with open(summary_file_name, 'r') as old_summary:
+          old_summary.readline()  # Skip header
+          new_summary_file_name = '{}.new'.format(summary_file_name)
+          with open(new_summary_file_name, 'w') as new_summary:
+            # Write new header
+            csv.writer(new_summary).writerow(self._summary_columns)
+            # Recover old data
+            new_summary.writelines(old_summary)
```

Comment:
Will this be valid CSV data? It seems like the lines from the old summary data will have fewer columns than the new header.